scripts/hid_configurator/configurator.py

```diff
@@ -263,38 +319,80 @@ def perform_dfu(dev, args):
     img_length = os.stat(dfu_image).st_size
     print('DFU image size: {} bytes'.format(img_length))
 
+    for i in range(3):
+        dfu_info = dfu_sync(dev, args)
+        if dfu_info is not None:
+            if dfu_info[0] != 0:
+                time.sleep(1)
+            else:
+                break
+
+    if dfu_info is None:
+        print('Cannot start DFU, device not responding')
+        return
+
+    if dfu_info[0] != 0:
+        print('Cannot start DFU, already in progress')
+        return
+
+    img_csum = file_crc(dfu_image)
+    if (dfu_info[1] == img_length) and (dfu_info[2] == img_csum) and (dfu_info[3] <= img_length):
+        print('Resume DFU at {}'.format(dfu_info[3]))
+        offset = dfu_info[3]
+    else:
+        offset = 0
+
+    dfu_start(dev, args, img_length, img_csum, offset)
+
     img_file = open(dfu_image, "rb")
-    offset = 0
+    img_file.seek(offset)
 
     try:
-        while True:
-            dfu_header = struct.pack('<II', img_length, offset)
-
-            chunk_data = img_file.read(EVENT_DATA_LEN_MAX - len(dfu_header))
+        while offset < img_length:
+            if offset % 4096 == 0:
+                success = False
+                dfu_info = dfu_sync(dev, args)
+                if dfu_info is None:
+                    break
+                if dfu_info[0] == 0:
+                    print('DFU interrupted by device')
+                    break
+                if (dfu_info[1] != img_length) or (dfu_info[2] != img_csum) or (dfu_info[3] != offset):
+                    print('Invalid sync information')
+                    break
+            chunk_data = img_file.read(EVENT_DATA_LEN_MAX)
             chunk_len = len(chunk_data)
             if chunk_len == 0:
                 break
 
             logging.debug('Send DFU request: offset {}, size {}'.format(offset, chunk_len))
 
-            event_data = dfu_header + chunk_data
-
             progress_bar(int(offset/img_length * 1000))
-            success = exchange_feature_report(dev, recipient, event_id, event_data, False)
+            success = exchange_feature_report(dev, recipient, event_id, chunk_data, False)
             if not success:
                 break
 
             offset += chunk_len
     except:
         success = False
-    finally:
-        img_file.close()
+
+    img_file.close()
+    print('')
 
     if success:
-        print('')
-        print('DFU transfer completed')
+        for i in range(3):
+            dfu_info = dfu_sync(dev, args)
+            if dfu_info is not None:
+                if dfu_info[0] != 0:
+                    time.sleep(1)
```

Comment:
Maybe comment why we need to sleep here?

Reply:
Done